### PR TITLE
Fix weapon effects (beams/railguns/bolts etc) interaction with obstacles

### PR DESCRIPTION
**Pull request by Starkku**
Prevents weapon effects like lasers, railgun particles, rad beams, electric bolts and waves from being drawn beyond where the projectile detonates (walls, solid buildings etc).

In addition particles spawned by `BehavesLike=Railgun` particle system as well as `AmbientDamage` are no longer cut off by elevation changes along trajectory.

## Summary by CodeRabbit

- **New Features**
	- Added new options for customizing fire particle systems and ambient damage effects.
	- Introduced the ability to show a briefing screen at the start of singleplayer missions.
	- Enhanced missile targeting, turret rotation, and jumpjet stability under EMP conditions.
	- Added customization options for disk laser radius and cluster scatter distance in projectiles.

- **Bug Fixes**
	- Fixed graphical issues with weapon effects not interacting correctly with obstacles.
	- Resolved problems with railgun and fire particles being cut off by elevation changes.
	- Corrected behavior of `AmbientDamage` with `IsRailgun=yes` to prevent cutoff by elevation.

- **Documentation**
	- Updated documentation to reflect new features and bug fixes.
	- Renamed the `Detonate.Warhead` property to `Detonate.WarheadType` in configuration files for clarity.

- **Refactor**
	- Removed outdated laser drawing logic to streamline weapon effect processing.

- **Chores**
	- Integrated new source files for the `ParticleSystemType` module.

### CODE REVIEW

**Comment by mevitar** [comment]
I managed to get this angle perfectly by accident and tested it with this build, and you can see here that the railgun is still drawing despite the projectile detonating on the edge of this solid building. Or, to be precise, on its corner (i know it can be hard to spot because of the animations i used. It was for IsDetachedRailgun=yes railgun, though, so don't know if it's supported.
railonedge

Or is it meant to work only for non-shp particles?

**Comment by Starkku** [comment]
> you can see here that the railgun is still drawing despite the projectile detonating on the edge of this solid building

Just to eliminate the most obvious cause of issues like this, this is `Inviso=true` projectile yes? The parity is not guaranteed in cases where it isn't and is actually technically impossible to do without a *massive* rewrite of all of the weapon/projectile logic. What is probably up for debate if the obstacle logic should be disabled if `Inviso` is not true.

> It was for IsDetachedRailgun=yes railgun, though, so don't know if it's supported.

Should be.

**Comment by mevitar** [comment]
> Just to eliminate the most obvious cause of issues like this, this is `Inviso=true` projectile yes?
it actually isn't. it's a straight trajectory projectile with an attached railgun. I didn't realize the fix will work only for Inviso=yes projectiles. Maybe that should be mentioned in the doc?

Retested it with an Inviso=yes projectile and that one does not even get blocked on this angle. It detonates on the target instead of the building, and the "beam" is also drawn from the firer to the target. From that, i can't say if the fix is working properly or not.

That pretty much means i'm out of ideas how to test this fix for solid buildings. Units will move around if they detect a building they can't fire through, and getting angles like this one is a matter of sheer luck.

**Comment by Starkku** [comment]
> it actually isn't. it's a straight trajectory projectile with an attached railgun. I didn't realize the fix will work only for Inviso=yes projectiles. Maybe that should be mentioned in the doc?

Documentation does have this line, can probably clarify if needed though.
```Fixed weapon effects like railguns, fire particles, lasers, beams, bolts and waves drawing beyond where Warhead detonates if it hits an obstacle like wall. Note that this is not necessarily same as the actual projectile travel path if it is not Inviso=true projectile.```

> That pretty much means i'm out of ideas how to test this fix for solid buildings. Units will move around if they detect a building they can't fire through, and getting angles like this one is a matter of sheer luck.

This is pretty much how it is since the obstacle logic affects rangefinding. The cases where the projectile is allowed to fire but is blocked by an obstacle are very situational but they do happen. It is usually hitting corners like in the example you had.

**Comment by mevitar** [comment]
> Documentation does have this line, can probably clarify if needed though.

Nah, that's enough. I thought it's not there because i couldn't find it. Apparently i was looking in the incorrect doc file (clicked on the big "View Docs" button instead of on the small "View docs" hyperlink text).

File: src/Ext/Techno/Hooks.WeaponEffects.cpp [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+}
+
+// Do not adjust map coordinates for railgun or fire stream particles that are below cell coordinates.
+DEFINE_HOOK(0x62B8BC, ParticleClass_CTOR_CoordAdjust, 0x6)
+{
+	enum { SkipCoordAdjust = 0x62B8CB };
+
+	GET(ParticleClass*, pThis, ESI);
+
+	if (pThis->ParticleSystem && (pThis->ParticleSystem->Type->BehavesLike == BehavesLike::Railgun || pThis->ParticleSystem->Type->BehavesLike == BehavesLike::Fire))
```
Comment (Metadorius) [inline src/Ext/Techno/Hooks.WeaponEffects.cpp:184]: Looks obnoxiously long.

**Review by Metadorius — COMMENTED** [review]
LGTM apart from comments

File: src/Ext/Techno/Hooks.WeaponEffects.cpp [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+	return 0;
+}
+
+// Restore original target values and unset obstacle cell.
+DEFINE_HOOK(0x6FF660, TechnoClass_FireAt_ObstacleCellUnset, 0x6)
+{
+	LEA_STACK(CoordStruct*, pTargetCoords, STACK_OFFSET(0xB0, -0x28));
+
+	auto coords = FireAtTemp::originalTargetCoords;
+	pTargetCoords = &coords;
```
Comment (chaserli) [inline src/Ext/Techno/Hooks.WeaponEffects.cpp:234]: Not sure what you were intending to do but nothing is done here. Were you trying to say `*pTargetCoords = coords`?

File: src/Ext/Techno/Hooks.WeaponEffects.cpp [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+	auto target = FireAtTemp::pOriginalTarget;
+
+	FireAtTemp::originalTargetCoords = CoordStruct::Empty;
+	FireAtTemp::pOriginalTarget = nullptr;
```
Comment (chaserli) [inline src/Ext/Techno/Hooks.WeaponEffects.cpp:238]: Some of these might not be indispensable,  those on ebp/esp should have remained intact

**Pull request merged**